go/database/mpt/verification_test.go

```diff
@@ -412,6 +415,116 @@ func TestVerification_ValueNodeHashModificationIsDetected(t *testing.T) {
 		}
 	})
 }
+
+func TestVerification_MissingCodeHashInCodeFileIsDetected(t *testing.T) {
+	runVerificationTest(t, func(t *testing.T, dir string, config MptConfig, roots []Root) {
+		if err := VerifyCodesAndForest(dir, config, roots, NilVerificationObserver{}); err == nil {
+			t.Errorf("missing hash in code file should have been detected")
+		}
+	})
+}
+
+func TestVerification_DifferentHashInCodeFileIsDetected(t *testing.T) {
+	runVerificationTest(t, func(t *testing.T, dir string, config MptConfig, roots []Root) {
+		testHash := common.Keccak256([]byte{1})
+		codes := map[common.Hash][]byte{
+			testHash: {2},
+		}
+		if err := writeCodes(codes, filepath.Join(dir, "codes.dat")); err != nil {
+			t.Fatalf("failed to write code file")
+		}
+
+		if err := VerifyCodesAndForest(dir, config, roots, NilVerificationObserver{}); err == nil {
+			t.Errorf("different hash in code file should have been detected")
+		}
+	})
+}
+
+func TestVerification_ExtraCodeHashInCodeFileIsDetected(t *testing.T) {
+	runVerificationTest(t, func(t *testing.T, dir string, config MptConfig, roots []Root) {
+		ctrl := gomock.NewController(t)
+		observer := NewMockVerificationObserver(ctrl)
+
+		testHash1 := common.Keccak256([]byte{1})
+		testHash2 := common.Keccak256([]byte{2})
+		codes := map[common.Hash][]byte{
+			testHash1: {1},
+			testHash2: {2},
+		}
+
+		gomock.InOrder(
+			observer.EXPECT().StartVerification(),
+			observer.EXPECT().Progress(WithSubstring("Obtaining read access to files")),
+			observer.EXPECT().Progress(WithSubstring("Checking contract codes")),
+			observer.EXPECT().Progress(WithSubstring(fmt.Sprintf("Contract %x is not referenced", testHash2))),
```

Comment:
Is not it that you are in reality looking just for this substring: "Contract %x is not referenced"?  Why mocking all other calls? I think the newly created matcher has a little effect, because the input are still full sentence, i.e. it solves only situation when a message is extended with some extra words. Is this likely?

Why not do it this simple way? 
```

		var triggered bool
		observer.EXPECT().StartVerification()
		observer.EXPECT().EndVerification(nil)
		observer.EXPECT().Progress(gomock.Any()).Do(func(msg string) {
			if strings.Contains(msg, fmt.Sprintf("Contract %x is not referenced", testHash2)) {
				triggered = true
			}
		}).AnyTimes()
...
                 if !triggered {
			t.Errorf("extra hash in code file should have been detected")
		}

```

I think it would serve the same purpose and its much simpler and easy to follow